{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to evolutionary simulations with Geonomics (Part 1)\n",
    "\n",
    "The goal of this notebook is to get you to think critically about how evolution happens across space. This notebook will also introduce you to the usefulness of modeling and simulations for understanding evolutionary scenarios. We are using a Python package call **\"Geonomics\"** that was conceived and created by Dr. Drew Terasaki Hart a former graduate student at UC Berkeley in Ian Wang's lab in the College of Natural Resources, Environmental Science, Policy and Management Department.\n",
    "\n",
    "* **Geographic variation**: As you've learned in lecture, there is often a geographic component to genetic divergence. Both genetic drift (founder effects in particular) and natural selection are facilitated by geographic isolation, while  gene flow tends to be impeded by it (although this is not always the case). It makes sense to simulate evolution in a programmatic way that allows for explicit and different spatial landscape scenarios, just like in real life.\n",
    "\n",
    "*Geonomics* allows the user to test hypotheses and run simulations about how the spatial arrangement and size of populations of organisms while incorporating any realistic or modeled geospatial **layers** of a landscape. The so-called forces of evolution can all be simulated: mutation, natural selection, founder effects, genetic drift, and gene-flow. The result is a very flexible platform to allow prediction of evolutionary outcomes during population divergence leading to speciation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Glossary\n",
    "\n",
    "**A glossary of key terms and concepts is available here [Geonomics Glossary Google Doc](https://docs.google.com/document/d/133njF4NFe2SH5zSASTnObL5wGRp3RalnkFM6Aih-Wf0/edit?usp=sharing)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Important notes: \n",
    "\n",
    "*Please don't be overwhelmed by the length of the code blocks in this notebook.* We intentionally left the code for you to see -- and exercises for you to modify the code -- so that you can think about the different parameters than need to be defined to construct a somewhat realistic evolutionary simulation.\n",
    "\n",
    "You need to read the **text cells** carefully and follow the instructions. Some **code cells** you will *run without changing*. Other times you will need to *edit the cells as instructed* to adjust the parameters of the model so we can simulate different scenarios. Reading the green `# comments` in Geonomics code will help to understand what you are actually doing during model set-up,  running of the model, and model results evaluation, and for **creating your own scenario** at the end. It is not necessary for you to read all of the `# comments` or code during this Notebook, this is what a user of the Geonomics package will refer to while running their simulations. We will also make use of **functions** which are blocks of code that start with `def` which perform a task (you do not need to edit or make your own functions); for example, we might use a function to plot a map or create a landscape."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helpful Reminders: \n",
    "\n",
    "### Jupyter notebooks\n",
    "When a cell is running the circle in the top right of the notebook next to Python 3 will be solid grey. Some cells take several seconds to run. Just be patient.\n",
    "\n",
    "\n",
    "### Text cells\n",
    "In a notebook, each rectangle containing text or code is called a *cell*.\n",
    "\n",
    "Text cells (like this one) can be edited by double-clicking on them. They're written in a simple format called [Markdown](http://daringfireball.net/projects/markdown/syntax) to add formatting and section headings.  You don't need to learn Markdown, but you might want to.\n",
    "\n",
    "After you edit a text cell, click the \"run cell\" button at the top to confirm any changes. (Try not to delete the instructions of the lab.)\n",
    "\n",
    "### Code cells\n",
    "Other cells contain code in the Python 3 language. Running a code cell will execute all of the code it contains.\n",
    "\n",
    "To run the code in a code cell, first click on that cell to activate it.  It'll be highlighted with a little green or blue rectangle.  Next, either press the Run button or hold down the `shift` key and press `return` or `enter`.\n",
    "\n",
    "The only code cells that need to be modified are right below a blue exercise box.\n",
    "\n",
    "### Comments\n",
    "Comments are statements in English that the computer ignores. We use comments to explain what the surrounding code does. Comments appear in green after the `#` symbol."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\" role=\"alert\" style=\"font-size:120%\">\n",
    "    \n",
    "**Agenda**:\n",
    "<ul>\n",
    "    <li>1. short intro about Geonomics package</li>\n",
    "    <li>2. see briefly how Geonomics works (create parameters file, edit, make model, run model, visualize)</li>\n",
    "    <li>3. run a two-stage simulation of neutral evolution</li>\n",
    "</ul>\n",
    "\n",
    "<p>\n",
    "\n",
    "</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>What does Geonomics do?</h2></b> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Geonomics is a Python package that simulates the genetic evolution of species distributed across landscapes. The landscapes can be arbitrarily complex, and can undergo environmental change over time. Species are composed of individuals, each of which has its own genome. The genes in those genomes can be used to determine the phenotypes of individuals, which can then undergo differential survival because of how poorly or well they are adapted to their local environments - that is, natural selection.\n",
    "\n",
    "The model operates by iterating through time steps (which you can think of being similar to generations). Each time step, a series of basic biological operations are carried out for all individuals. Over time, patterns of genetic diversity can build up on the landscape. These are what we are interested in studying, because these patterns in real-world species can help us make inferences about important things that have happened or are happening in nature, such as migration, adaptation, extinction, and the like.  In the real world, observing these kinds of changes could take thousands of years (depending on the organism), so models help us see these patterns in a much shorter time frame!\n",
    "\n",
    "Below are a conceptual diagram and its caption, taken from [the Geonomics homepage](https://github.com/drewhart/geonomics). They give a simple explanation of how each time step works."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![conceptual_diagram](https://github.com/erthward/geonomics/raw/main/img/conceptual_diagram.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the center is a species on a multi-layer landscape that includes a selection layer (above) and a movement and carrying capacity layer (below). Surrounding the landscape is a flow-diagram of the major operations during a time step.\n",
    "\n",
    "- **movement**: During the movement stages (top-left), individuals move along movement vectors drawn from various distribution options.\n",
    "- **mating**: During the mating stage (top-right), an individual (purple outline) randomly chooses a mate (green outline) from all potential mates within its mating radius (dashed circle). The resulting offspring (dashed outline) disperses from its parents' midpoint along a randomly drawn dispersal vector.\n",
    "- **mortality**: During the mortality stage (bottom-right), deaths are modeled as a Bernoulli process, with the probability of mortality being a product of density-dependence and selection on all traits.\n",
    "- **changes**: During this stage (bottom-left), demographic change events (not pictured) and environmental change events (represented as a series of change rasters corresponding to scheduled time steps, t1, t2, â€¦, tn), take place.\n",
    "\n",
    "Let's think about this in terms of the real world.  Think of the landscape in Geonomics as an island that we can control.  The different layers represent different environmental conditions on that island, like elevation, temperature, rainfall, food availability, etc.  Right now, there is one species on our island, and each time step represents one generation of individuals.  We will use mice (like rock pocket mice) for our example.\n",
    "\n",
    "During each time step, our mice will do three different things: move around (movement), find mates and produce offspring (mating), and die (mortality). The Geonomics model allows us to dictate how our mice do each of these things- for example, we can say that a mouse can only move within 15 meters of where it was born and can only find a mate within one meter of its location.  We can also specify how long each mouse lives, and if certain environmental conditions (like high temperature) will increase their likelihood of death or survival.\n",
    "\n",
    "While our mice are doing their thing, our landscape model will also change based on any predetermined parameters (changes).  Before running the model, we can set up landscape events to occur at certain time steps.  For example, we can say that for every 10 time steps, the temperature will increase by one degree Celsius.  Since we can link environmental factors to mortality, this will let us model how increasing temperature will affect our mice population over hundreds of generations!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Okay! Now on to the code!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>Import needed packages</h2></b> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you are on your own computer and using Anaconda, run the following cell to install the required python packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!conda config --add channels conda-forge\n",
    "!conda install msprime -y\n",
    "!conda install geopandas -y\n",
    "!conda install rasterio -y\n",
    "!conda install bitarray -y\n",
    "!pip install NLMpy -q\n",
    "!pip install geonomics -q"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You only need to do this once."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you are using JupyterHub, run the following cell to install the required python packages. DO NOT run this cell if you are using Anaconda on your own computer:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install msprime\n",
    "!pip install numba\n",
    "!pip install NLMpy\n",
    "!pip install bitarray\n",
    "!pip install geonomics==1.4.6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the following cell to bring in the necessary Python packages for our exercise today. Recall that you need to run this cell and any cells above the one you are working on each time you relaunch the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import geonomics as gnx\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# and also set Matplotlib's default plotting style and plot size\n",
    "%matplotlib inline\n",
    "plt.rcParams[\"figure.figsize\"] = (9,4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>Define some helper functions</h2></b> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we will define some functions to create simple, geometric arrays, which we'll use as landscapes for our simulations.\n",
    "\n",
    "__No need to alter the code here.__ Just run it. We'll use the functions below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################\n",
    "# RUN THIS CODE WITHOUT CHANGING\n",
    "################################\n",
    "\n",
    "def make_unif_array(n):\n",
    "    \"\"\"Makes a square array of ones, size n x n cells.\"\"\"\n",
    "    array = np.ones((n,n))\n",
    "    return array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################\n",
    "# RUN THIS CODE WITHOUT CHANGING\n",
    "################################\n",
    "\n",
    "def add_vert_barrier(array, width = 5):\n",
    "    \"\"\"Returns a new array, which is the input array but with an added vertical barrier of the defined width.\"\"\"\n",
    "    if array.shape[0] % 2 != width % 2:\n",
    "        width += 1\n",
    "    assert width <= array.shape[0] - 2, (\"The width of the barrier should be \"\n",
    "                                       \"at least 2 less than the width of the landscape.\")\n",
    "    left_extent = int((array.shape[0] - width)/2)\n",
    "    right_extent = left_extent + width\n",
    "    new_array = np.copy(array)\n",
    "    new_array[:, left_extent:right_extent] = 0\n",
    "    return new_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################\n",
    "# RUN THIS CODE WITHOUT CHANGING\n",
    "################################\n",
    "\n",
    "def make_horz_grad_array(n, grad_width=8):\n",
    "    \"\"\"Makes a square array with a horizontal gradient\n",
    "    running from 0 to 1, size n x n cells.\"\"\"\n",
    "    assert grad_width <= n, \"Argument 'grad_width' must be <= argument 'n'.\"\n",
    "    if n % 2 != grad_width % 2:\n",
    "        grad_width += 1\n",
    "    grad_vals = np.linspace(0, 1, grad_width)\n",
    "    append_len = int((n - grad_width) / 2)\n",
    "    grad_vals = np.hstack([np.zeros((append_len)), grad_vals, np.ones((append_len))])\n",
    "    array = np.vstack([grad_vals.T]*n)\n",
    "    return array"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll also load some functions that will use the multivariate statistical method called Principal Component Analysis (PCA) to assign genetic-relatedness values to all the individuals in a species. \n",
    "\n",
    "The first function will plot all individuals in \"genetic relatedness space\", such that distances between individuals represent how genetically different they are. \n",
    "\n",
    "The second function will map individuals on the landscape, colored by their genetic relatedness, such that more similar-colored individuals are more related.\n",
    "\n",
    "We can use these visualizations to get a sense for how the genetic relatedness of the individuals in our species changes over time, because of distance, landscape features, population size and density, selective pressure, and other influences.\n",
    "\n",
    "__No need to alter the code here.__ Just run it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################\n",
    "# RUN THIS CODE WITHOUT CHANGING\n",
    "################################\n",
    "\n",
    "# function for running and plotting genetic PCA\n",
    "def plot_PCA(mod):\n",
    "    from copy import deepcopy\n",
    "    from sklearn.decomposition import PCA\n",
    "    figsize = 6\n",
    "    species = mod.comm[0]\n",
    "    # get array of resulting genomic data (i.e. 'speciome'),\n",
    "    # genotypes averaged by individual\n",
    "    speciome = np.mean(species._get_genotypes(), axis=2)\n",
    "    # run PCA on speciome\n",
    "    pca = PCA(n_components=2)\n",
    "    PCs = pca.fit_transform(speciome)\n",
    "    # normalize the PC results\n",
    "    norm_PCs = (PCs - np.min(PCs,\n",
    "                             axis=0)) / (np.max(PCs,\n",
    "                                                axis=0) - np.min(PCs,\n",
    "                                                                 axis=0))\n",
    "    # assign a value to each species, 0 or 1, indicating whether they're located on\n",
    "    # the left or right vertical half of the landscape\n",
    "    ind_colors = ['#00ffff' if ind.x < mod.land.dim[0]/2 else '#ff00ff' for ind in species.values()]\n",
    "    # plot individuals on PCs 1 and 2, colored by their landscape half\n",
    "    fig = plt.figure(figsize=(figsize, figsize), dpi= 80, facecolor='w', edgecolor='k')\n",
    "    plt.scatter(norm_PCs[:,0], norm_PCs[:, 1], color = ind_colors)\n",
    "    plt.xlabel('genetic PC 1')\n",
    "    plt.ylabel('genetic PC 2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################\n",
    "# RUN THIS CODE WITHOUT CHANGING\n",
    "################################\n",
    "\n",
    "# function for running and plotting genetic PCA\n",
    "def map_PCA(mod, lyr_num=0, mask=True):\n",
    "    from copy import deepcopy\n",
    "    from sklearn.decomposition import PCA\n",
    "    cmaps = {0: plt.cm.RdBu.copy(), 1: plt.cm.BrBG_r.copy()}\n",
    "    mark_size = 60\n",
    "    figsize = 8\n",
    "    species = mod.comm[0]\n",
    "    land = mod.land\n",
    "    # get array of resulting genomic data (i.e. 'speciome'),\n",
    "    # genotypes meaned by individual\n",
    "    speciome = np.mean(species._get_genotypes(), axis=2)\n",
    "    # run PCA on speciome\n",
    "    pca = PCA(n_components=3)\n",
    "    PCs = pca.fit_transform(speciome)\n",
    "    # normalize the PC results\n",
    "    norm_PCs = (PCs - np.min(PCs,\n",
    "                             axis=0)) / (np.max(PCs,\n",
    "                                                axis=0) - np.min(PCs,\n",
    "                                                                 axis=0))\n",
    "    # use first 3 PCs to get normalized values for R, G, & B colors\n",
    "    PC_colors = norm_PCs * 255\n",
    "    # scatter all individuals on top of landscape, colored by the\n",
    "    # RBG colors developed from the first 3 geonmic PCs\n",
    "    xs = mod.comm[0]._get_x()\n",
    "    ys = mod.comm[0]._get_y()\n",
    "    # get environmental raster, with barrier masked out\n",
    "    env = deepcopy(mod.land[lyr_num].rast)\n",
    "    if mask:\n",
    "        env[mod.land[lyr_num].rast == 0] = np.nan\n",
    "    # create light colormap for plotting landscape\n",
    "    # bot = plt.cm.get_cmap('Blues', 256)(np.linspace(0.4, 0.45, 2))[0]\n",
    "    # top = plt.cm.get_cmap('Reds', 256)(np.linspace(0.4, 0.45, 2))[0]\n",
    "    # cols = np.vstack((top, bot))\n",
    "    # cmap = mpl.colors.ListedColormap(cols, name='OrangeBlue')\n",
    "    cmap = cmaps[lyr_num]\n",
    "    cmap.set_bad(color='#8C8C8C')\n",
    "    # plot landscape\n",
    "    fig = plt.figure(figsize=(figsize, figsize), dpi= 80, facecolor='w', edgecolor='k')\n",
    "    # plt.imshow(masked_env, cmap=cmap, alpha=0.8)\n",
    "    plt.pcolormesh(land._x_cell_bds, land._y_cell_bds, env, cmap=cmap)\n",
    "    # scatter plot of individuals, colored by composite PC score\n",
    "    plt.scatter(xs, ys, c=PC_colors/255.0, s=mark_size, edgecolors='black')\n",
    "    # fix x and y limits\n",
    "    [f([dim - 0.5 for dim in (0, mod.land.dim[0])]) for f in (plt.xlim,\n",
    "                                                              plt.ylim)]\n",
    "    # get rid of x and y ticks\n",
    "    [f([]) for f in (plt.xticks, plt.yticks)]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-danger\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>PART 1: Neutral evolution</h2></b> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following section will walk through a neutral evolutionary simulation on a very simple landscape.\n",
    "\n",
    "This simulation will highlight the effects of geographic and cost distance on genetic relatedness.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>Set the parameter values </h2></b> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following long block of code creates a Geonomics parameters object.\n",
    "\n",
    "That object in turn will be used to instantiate (i.e. make a new, independent example of) a Geonomics model, which we can use to run simulations.\n",
    "\n",
    "The model will be composed of:\n",
    "\n",
    "- **a landscape**, with one **layer** that we'll use as both the environment and the carrying-capacity map (i.e. our island)\n",
    "- **a species**, which moves around on the landscape and has genomes containing only neutral loci (i.e. our mice)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# neutral_demo.py\n",
    "\n",
    "# This is a parameters file generated by Geonomics\n",
    "# (by the gnx.make_parameters_file() function).\n",
    "\n",
    "\n",
    "                   #   ::::::          :::    :: :::::::::::#\n",
    "             #::::::    ::::   :::      ::    :: :: ::::::::::: ::#\n",
    "          #:::::::::     ::            ::   :::::::::::::::::::::::::#\n",
    "        #::::::::::                      :::::::::: :::::: ::::::::  ::#\n",
    "      #  : ::::  ::                    ::::  : ::    :::::::: : ::  :    #\n",
    "     # GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #\n",
    "    # GG     EE    OO   OO  NNN  NN  OO   OO  MM   MM   II   CC     SS     #\n",
    "    # GG     EE   OO     OO NN N NN OO     OO MMM MMM   II   CC     SSSSSS #\n",
    "    # GG GGG EEEE OO     OO NN  NNN OO     OO MM M MM   II   CC         SS #\n",
    "    # GG   G EE    OO   OO  NN   NN  OO   OO  MM   MM   II   CC        SSS #\n",
    "     # GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #\n",
    "      #    : ::::::::               :::::::::: ::              ::  :   : #\n",
    "        #:    :::::                    :::::: :::             :::::::  #\n",
    "          #    :::                      :::::  ::              ::::: #\n",
    "             #  ::                      ::::                      #\n",
    "                   #                                        #\n",
    "                      #  :: ::    :::             #\n",
    "\n",
    "\n",
    "params = {\n",
    "###############################################################################\n",
    "\n",
    "###################\n",
    "#### LANDSCAPE ####\n",
    "###################\n",
    "    'landscape': {\n",
    "\n",
    "    ##############\n",
    "    #### main ####\n",
    "    ##############\n",
    "        'main': {\n",
    "            #x,y (a.k.a. j,i) dimensions of the Landscape\n",
    "            'dim':                      (20,20),\n",
    "            #x,y resolution of the Landscape\n",
    "            'res':                      (1,1),\n",
    "            #x,y coords of upper-left corner of the Landscape\n",
    "            'ulc':                      (0,0),\n",
    "            #projection of the Landscape\n",
    "            'prj':                      None,\n",
    "            }, # <END> 'main'\n",
    "\n",
    "    ################\n",
    "    #### layers ####\n",
    "    ################\n",
    "        'layers': {\n",
    "\n",
    "            #layer name (LAYER NAMES MUST BE UNIQUE!)\n",
    "            'lyr_0': {\n",
    "\n",
    "        #######################################\n",
    "        #### layer num. 0: init parameters ####\n",
    "        #######################################\n",
    "\n",
    "                #initiating parameters for this layer\n",
    "                'init': {\n",
    "\n",
    "                    #parameters for a 'defined'-type Layer\n",
    "                    'defined': {\n",
    "                        #raster to use for the Layer\n",
    "                        'rast':                   np.ones((20,20)),\n",
    "                        #point coordinates\n",
    "                        'pts':                    None,\n",
    "                        #point values\n",
    "                        'vals':                   None,\n",
    "                        #interpolation method {None, 'linear', 'cubic',\n",
    "                        #'nearest'}\n",
    "                        'interp_method':          None,\n",
    "\n",
    "                        }, # <END> 'defined'\n",
    "\n",
    "                    }, # <END> 'init'\n",
    "\n",
    "                }, # <END> layer num. 0\n",
    "\n",
    "\n",
    "\n",
    "    #### NOTE: Individual Layers' sections can be copy-and-pasted (and\n",
    "    #### assigned distinct keys and names), to create additional Layers.\n",
    "\n",
    "\n",
    "            } # <END> 'layers'\n",
    "\n",
    "        }, # <END> 'landscape'\n",
    "\n",
    "\n",
    "###############################################################################\n",
    "\n",
    "###################\n",
    "#### COMMUNITY ####\n",
    "###################\n",
    "    'comm': {\n",
    "\n",
    "        'species': {\n",
    "\n",
    "            #species name (SPECIES NAMES MUST BE UNIQUE!)\n",
    "            'spp_0': {\n",
    "\n",
    "            #####################################\n",
    "            #### spp num. 0: init parameters ####\n",
    "            #####################################\n",
    "\n",
    "                'init': {\n",
    "                    #starting number of individs\n",
    "                    'N':                250,\n",
    "                    #carrying-capacity Layer name\n",
    "                    'K_layer':          'lyr_0',\n",
    "                    #multiplicative factor for carrying-capacity layer\n",
    "                    'K_factor':         1,\n",
    "                    }, # <END> 'init'\n",
    "\n",
    "            #######################################\n",
    "            #### spp num. 0: mating parameters ####\n",
    "            #######################################\n",
    "\n",
    "                'mating'    : {\n",
    "                    #age(s) at sexual maturity (if tuple, female first)\n",
    "                    'repro_age':                0,\n",
    "                    #whether to assign sexes\n",
    "                    'sex':                      False,\n",
    "                    #ratio of males to females\n",
    "                    'sex_ratio':                1/1,\n",
    "                    #whether P(birth) should be weighted by parental dist\n",
    "                    'dist_weighted_birth':       False,\n",
    "                    #intrinsic growth rate\n",
    "                    'R':                        0.5,\n",
    "                    #intrinsic birth rate (MUST BE 0<=b<=1)\n",
    "                    'b':                        0.2,\n",
    "                    #expectation of distr of n offspring per mating pair\n",
    "                    'n_births_distr_lambda':    1,\n",
    "                    #whether n births should be fixed at n_births_dist_lambda\n",
    "                    'n_births_fixed':           True,\n",
    "                    #radius of mate-search area\n",
    "                    'mating_radius':            10,\n",
    "                    #whether individs should choose nearest neighs as mates\n",
    "                    'choose_nearest_mate':        False,\n",
    "                    #whether mate-choice should be inverse distance-weighted                                                                                                                  \n",
    "                    'inverse_dist_mating':      False,\n",
    "                    }, # <END> 'mating'\n",
    "\n",
    "            ##########################################\n",
    "            #### spp num. 0: mortality parameters ####\n",
    "            ##########################################\n",
    "\n",
    "                'mortality'     : {\n",
    "                    #maximum age\n",
    "                    'max_age':                      None,\n",
    "                    #min P(death) (MUST BE 0<=d_min<=1)\n",
    "                    'd_min':                        0,\n",
    "                    #max P(death) (MUST BE 0<=d_max<=1)\n",
    "                    'd_max':                        1,\n",
    "                    #width of window used to estimate local pop density\n",
    "                    'density_grid_window_width':    None,\n",
    "                    }, # <END> 'mortality'\n",
    "\n",
    "            #########################################\n",
    "            #### spp num. 0: movement parameters ####\n",
    "            #########################################\n",
    "\n",
    "                'movement': {\n",
    "                    #whether or not the species is mobile\n",
    "                    'move':                                 True,\n",
    "                    #mode of distr of movement direction\n",
    "                    'direction_distr_mu':                   0,\n",
    "                    #concentration of distr of movement direction\n",
    "                    'direction_distr_kappa':                0,\n",
    "                    #1st param of distr of movement distance\n",
    "                    'movement_distance_distr_param1':       0.5,\n",
    "                    #2nd param of distr of movement distance\n",
    "                    'movement_distance_distr_param2':       5e-8,\n",
    "                    #movement distance distr to use ('levy' or 'wald')\n",
    "                    'movement_distance_distr':              'levy',\n",
    "                    #1st param of distr of dispersal distance\n",
    "                    'dispersal_distance_distr_param1':      0,\n",
    "                    #2nd param of distr of dispersal distance\n",
    "                    'dispersal_distance_distr_param2':      5e-14,\n",
    "                    #dispersal distance distr to use ('levy' or 'wald')\n",
    "                    'dispersal_distance_distr':             'levy',\n",
    "                    'move_surf'     : {\n",
    "                        #move-surf Layer name\n",
    "                        'layer':                'lyr_0',\n",
    "                        #whether to use mixture distrs\n",
    "                        'mixture':              True,\n",
    "                        #concentration of distrs\n",
    "                        'vm_distr_kappa':       12,\n",
    "                        #length of approximation vectors for distrs\n",
    "                        'approx_len':           5000,\n",
    "                        }, # <END> 'move_surf'\n",
    "\n",
    "                    },    # <END> 'movement'\n",
    "\n",
    "\n",
    "            #####################################################\n",
    "            #### spp num. 0: genomic architecture parameters ####\n",
    "            #####################################################\n",
    "\n",
    "                'gen_arch': {\n",
    "                    #file defining custom genomic arch\n",
    "                    'gen_arch_file':            None,\n",
    "                    #num of loci\n",
    "                    'L':                        100,\n",
    "                    #starting allele frequency (None to draw freqs randomly)\n",
    "                    'start_p_fixed':            0.5,\n",
    "                    #whether to start neutral locus freqs at 0\n",
    "                    'start_neut_zero':          False,\n",
    "                    #genome-wide per-base neutral mut rate (0 to disable)\n",
    "                    'mu_neut':                  0,\n",
    "                    #genome-wide per-base deleterious mut rate (0 to disable)\n",
    "                    'mu_delet':                 0,\n",
    "                    #shape of distr of deleterious effect sizes\n",
    "                    'delet_alpha_distr_shape':  0.2,\n",
    "                    #scale of distr of deleterious effect sizes\n",
    "                    'delet_alpha_distr_scale':  0.2,\n",
    "                    #alpha of distr of recomb rates\n",
    "                    'r_distr_alpha':            None,\n",
    "                    #beta of distr of recomb rates\n",
    "                    'r_distr_beta':             None,\n",
    "                    #whether loci should be dominant (for allele '1')\n",
    "                    'dom':                      False,\n",
    "                    #whether to allow pleiotropy\n",
    "                    'pleiotropy':               False,\n",
    "                    #custom fn for drawing recomb rates\n",
    "                    'recomb_rate_custom_fn':    None,\n",
    "                    #number of recomb paths to hold in memory\n",
    "                    'n_recomb_paths_mem':       int(1e4),\n",
    "                    #total number of recomb paths to simulate\n",
    "                    'n_recomb_paths_tot':       int(1e5),\n",
    "                    #num of crossing-over events (i.e. recombs) to simulate\n",
    "                    'n_recomb_sims':            10_000,\n",
    "                    #whether to jitter recomb bps, to correctly track num_trees\n",
    "                    'jitter_breakpoints':       False,\n",
    "                    #whether to generate recombination paths at each timestep\n",
    "                    'allow_ad_hoc_recomb':      False,\n",
    "                    #whether to use tskit (to record full spatial pedigree)\n",
    "                    'use_tskit':                True,\n",
    "                    #time step interval for simplication of tskit tables\n",
    "                    'tskit_simp_interval':      100,\n",
    "                    #whether to save mutation logs\n",
    "                    'mut_log':                  False,\n",
    "\n",
    "                    }, # <END> 'gen_arch'\n",
    "\n",
    "\n",
    "                }, # <END> spp num. 0\n",
    "\n",
    "\n",
    "\n",
    "    #### NOTE: individual Species' sections can be copy-and-pasted (and\n",
    "    #### assigned distinct keys and names), to create additional Species.\n",
    "\n",
    "\n",
    "            }, # <END> 'species'\n",
    "\n",
    "        }, # <END> 'comm'\n",
    "\n",
    "\n",
    "###############################################################################\n",
    "\n",
    "###############\n",
    "#### MODEL ####\n",
    "###############\n",
    "    'model': {\n",
    "        #total Model runtime (in timesteps)\n",
    "        'T':            100,\n",
    "        #min burn-in runtime (in timesteps)\n",
    "        'burn_T':       30,\n",
    "        #seed number\n",
    "        'num':          None,                                      \n",
    "\n",
    "        ###############################\n",
    "        #### iterations parameters ####\n",
    "        ###############################\n",
    "        'its': {\n",
    "            #num iterations\n",
    "            'n_its':            1,\n",
    "            #whether to randomize Landscape each iteration\n",
    "            'rand_landscape':   False,\n",
    "            #whether to randomize Community each iteration\n",
    "            'rand_comm':        False,\n",
    "            #whether to burn in each iteration\n",
    "            'repeat_burn':      False,\n",
    "            #whether to randomize GenomicArchitectures each iteration\n",
    "            'rand_genarch':     True,              \n",
    "            }, # <END> 'iterations'\n",
    "\n",
    "\n",
    "\n",
    "        } # <END> 'model'\n",
    "\n",
    "    } # <END> params\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>Create a model from those parameters</h2></b> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make our params dict into a proper Geonomics ParamsDict object\n",
    "params = gnx.make_params_dict(params, 'neutral_demo')\n",
    "#then use it to make a model\n",
    "mod = gnx.make_model(parameters=params, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#take a look at the resulting object\n",
    "mod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot the resulting object\n",
    "#spp = species, in this case we just have one species (species 0)\n",
    "#lyr = landscape layer, in this case we only have one layer (layer 0)\n",
    "mod.plot(spp=0, lyr=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Concept check**: *What are the black dots on the gray background?*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>Run the model</h2></b> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Geonomics allows us to use our model in either of two ways:\n",
    "\n",
    "- **'run'**: Run the model to completion (i.e. for the number of iterations and the number of timesteps per iteration that are stipulated in the model's parameters). This is more convenient for leaving a model running and generating data.\n",
    "\n",
    "- **'walk'**: Run the model for any number of timesteps, then stop it. This is more convenient for running a model interactively and inspecting it. This is what we'll use.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#First we have to run the model burn-in\n",
    "\n",
    "#T=10000 allows the model run up to 10000 steps, but it will burn in much faster than that.\n",
    "\n",
    "#After it runs, scroll to the bottom to see how the time steps progressed.\n",
    "\n",
    "mod.walk(T=10000, mode='burn')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Concept check** *How many individuals did you start and end with? How many births and deaths were there at the beginning / and births and deaths during the last time step?*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now we can plot it again to see the burned-in stable state.\n",
    "# note, mod.plot(0, 0) is the same as mod.plot(spp=0, lyr=0)\n",
    "mod.plot(0, 0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now we can run the main phase of the model for any desired number of timesteps, then plot it again.\n",
    "mod.walk(10, 'main')\n",
    "mod.plot(0, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "### **Concept check** *As you run the main phase of the model, and plot the result, what is changing?*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>Make the landscape heterogeneous</h2></b> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**What if we make the same model, but use a more interesting landscape?**\n",
    "\n",
    "Here we'll repeat the same process, but make our landscape using a **neutral landscape model (NLM)** algorithm.  We do this by changing some of the parameters in the 'lyr_0' section; you can compare that to the parameters code above. The NLM will produce a spatially autocorrelated pattern - i.e. one in which the landscape variable varies across space, but in a way that nearby values are more similar than distant values.  For example, if our variable is ground color, areas close to one another will be closer in color (i.e. dark spaces will border other dark spaces).\n",
    "\n",
    "Landscape ecologists often use NLMs as null models against which real-world patterns can be compared, to draw inferences about what might have generated real-world patterns. NLMs do not represent actual landscapes, but serve as the standard against which actual landscapes may be compared. This is referred to as a **null model**.\n",
    "\n",
    "*Here we won't use the NLM as a null model for hypothesis testing. Instead, we'll just use it as a quick way to get a more interesting landscape with a realistic pattern.* In this case, we will be using an NLM to create a **carrying capacity** layer where cells with higher values can support more individuals."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# neutral_demo_NLM.py\n",
    "\n",
    "# This is a parameters file generated by Geonomics\n",
    "# (by the gnx.make_parameters_file() function).\n",
    "\n",
    "\n",
    "                   #   ::::::          :::    :: :::::::::::#\n",
    "             #::::::    ::::   :::      ::    :: :: ::::::::::: ::#\n",
    "          #:::::::::     ::            ::   :::::::::::::::::::::::::#\n",
    "        #::::::::::                      :::::::::: :::::: ::::::::  ::#\n",
    "      #  : ::::  ::                    ::::  : ::    :::::::: : ::  :    #\n",
    "     # GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #\n",
    "    # GG     EE    OO   OO  NNN  NN  OO   OO  MM   MM   II   CC     SS     #\n",
    "    # GG     EE   OO     OO NN N NN OO     OO MMM MMM   II   CC     SSSSSS #\n",
    "    # GG GGG EEEE OO     OO NN  NNN OO     OO MM M MM   II   CC         SS #\n",
    "    # GG   G EE    OO   OO  NN   NN  OO   OO  MM   MM   II   CC        SSS #\n",
    "     # GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #\n",
    "      #    : ::::::::               :::::::::: ::              ::  :   : #\n",
    "        #:    :::::                    :::::: :::             :::::::  #\n",
    "          #    :::                      :::::  ::              ::::: #\n",
    "             #  ::                      ::::                      #\n",
    "                   #                                        #\n",
    "                      #  :: ::    :::             #\n",
    "\n",
    "\n",
    "params = {\n",
    "###############################################################################\n",
    "\n",
    "###################\n",
    "#### LANDSCAPE ####\n",
    "###################\n",
    "    'landscape': {\n",
    "\n",
    "    ##############\n",
    "    #### main ####\n",
    "    ##############\n",
    "        'main': {\n",
    "            #x,y (a.k.a. j,i) dimensions of the Landscape\n",
    "            'dim':                      (20,20),\n",
    "            #x,y resolution of the Landscape\n",
    "            'res':                      (1,1),\n",
    "            #x,y coords of upper-left corner of the Landscape\n",
    "            'ulc':                      (0,0),\n",
    "            #projection of the Landscape\n",
    "            'prj':                      None,\n",
    "            }, # <END> 'main'\n",
    "\n",
    "    ################\n",
    "    #### layers ####\n",
    "    ################\n",
    "        'layers': {\n",
    "\n",
    "            #layer name (LAYER NAMES MUST BE UNIQUE!)\n",
    "            'lyr_0': {\n",
    "\n",
    "        #######################################\n",
    "        #### layer num. 0: init parameters ####\n",
    "        #######################################\n",
    "\n",
    "                #initiating parameters for this layer\n",
    "                'init': {\n",
    "                    \n",
    "                    #parameters for an 'nlmpy'-type Layer\n",
    "                    'nlmpy': {\n",
    "                        #nlmpy function to use the create this Layer\n",
    "                        'function':                 'mpd',\n",
    "                        #number of rows (MUST EQUAL LANDSCAPE DIMENSION y!)\n",
    "                        'nRow':                     20,\n",
    "                        #number of cols (MUST EQUAL LANDSCAPE DIMENSION x!)\n",
    "                        'nCol':                     20,\n",
    "                        #level of spatial autocorrelation in element values\n",
    "                        'h':                        1,\n",
    "\n",
    "                        }, # <END> 'nlmpy'\n",
    "\n",
    "                    }, # <END> 'init'\n",
    "\n",
    "                }, # <END> layer num. 0\n",
    "\n",
    "\n",
    "\n",
    "    #### NOTE: Individual Layers' sections can be copy-and-pasted (and\n",
    "    #### assigned distinct keys and names), to create additional Layers.\n",
    "\n",
    "\n",
    "            } # <END> 'layers'\n",
    "\n",
    "        }, # <END> 'landscape'\n",
    "\n",
    "\n",
    "###############################################################################\n",
    "\n",
    "###################\n",
    "#### COMMUNITY ####\n",
    "###################\n",
    "    'comm': {\n",
    "\n",
    "        'species': {\n",
    "\n",
    "            #species name (SPECIES NAMES MUST BE UNIQUE!)\n",
    "            'spp_0': {\n",
    "\n",
    "            #####################################\n",
    "            #### spp num. 0: init parameters ####\n",
    "            #####################################\n",
    "\n",
    "                'init': {\n",
    "                    #starting number of individs\n",
    "                    'N':                250,\n",
    "                    #carrying-capacity Layer name\n",
    "                    'K_layer':          'lyr_0',\n",
    "                    #multiplicative factor for carrying-capacity layer\n",
    "                    'K_factor':         1,\n",
    "                    }, # <END> 'init'\n",
    "\n",
    "            #######################################\n",
    "            #### spp num. 0: mating parameters ####\n",
    "            #######################################\n",
    "\n",
    "                'mating'    : {\n",
    "                    #age(s) at sexual maturity (if tuple, female first)\n",
    "                    'repro_age':                0,\n",
    "                    #whether to assign sexes\n",
    "                    'sex':                      False,\n",
    "                    #ratio of males to females\n",
    "                    'sex_ratio':                1/1,\n",
    "                    #whether P(birth) should be weighted by parental dist\n",
    "                    'dist_weighted_birth':       False,\n",
    "                    #intrinsic growth rate\n",
    "                    'R':                        0.5,\n",
    "                    #intrinsic birth rate (MUST BE 0<=b<=1)\n",
    "                    'b':                        0.2,\n",
    "                    #expectation of distr of n offspring per mating pair\n",
    "                    'n_births_distr_lambda':    1,\n",
    "                    #whether n births should be fixed at n_births_dist_lambda\n",
    "                    'n_births_fixed':           True,\n",
    "                    #radius of mate-search area\n",
    "                    'mating_radius':            10,\n",
    "                    #whether individs should choose nearest neighs as mates\n",
    "                    'choose_nearest_mate':        False,\n",
    "                    #whether mate-choice should be inverse distance-weighted                                                                                                                  \n",
    "                    'inverse_dist_mating':      False,\n",
    "                    }, # <END> 'mating'\n",
    "\n",
    "            ##########################################\n",
    "            #### spp num. 0: mortality parameters ####\n",
    "            ##########################################\n",
    "\n",
    "                'mortality'     : {\n",
    "                    #maximum age\n",
    "                    'max_age':                      None,\n",
    "                    #min P(death) (MUST BE 0<=d_min<=1)\n",
    "                    'd_min':                        0,\n",
    "                    #max P(death) (MUST BE 0<=d_max<=1)\n",
    "                    'd_max':                        1,\n",
    "                    #width of window used to estimate local pop density\n",
    "                    'density_grid_window_width':    None,\n",
    "                    }, # <END> 'mortality'\n",
    "\n",
    "            #########################################\n",
    "            #### spp num. 0: movement parameters ####\n",
    "            #########################################\n",
    "\n",
    "                'movement': {\n",
    "                    #whether or not the species is mobile\n",
    "                    'move':                                 True,\n",
    "                    #mode of distr of movement direction\n",
    "                    'direction_distr_mu':                   0,\n",
    "                    #concentration of distr of movement direction\n",
    "                    'direction_distr_kappa':                0,\n",
    "                    #1st param of distr of movement distance\n",
    "                    'movement_distance_distr_param1':       0.5,\n",
    "                    #2nd param of distr of movement distance\n",
    "                    'movement_distance_distr_param2':       5e-8,\n",
    "                    #movement distance distr to use ('levy' or 'wald')\n",
    "                    'movement_distance_distr':              'levy',\n",
    "                    #1st param of distr of dispersal distance\n",
    "                    'dispersal_distance_distr_param1':      0,\n",
    "                    #2nd param of distr of dispersal distance\n",
    "                    'dispersal_distance_distr_param2':      5e-14,\n",
    "                    #dispersal distance distr to use ('levy' or 'wald')\n",
    "                    'dispersal_distance_distr':             'levy',\n",
    "                    'move_surf'     : {\n",
    "                        #move-surf Layer name\n",
    "                        'layer':                'lyr_0',\n",
    "                        #whether to use mixture distrs\n",
    "                        'mixture':              True,\n",
    "                        #concentration of distrs\n",
    "                        'vm_distr_kappa':       12,\n",
    "                        #length of approximation vectors for distrs\n",
    "                        'approx_len':           5000,\n",
    "                        }, # <END> 'move_surf'\n",
    "\n",
    "                    },    # <END> 'movement'\n",
    "\n",
    "\n",
    "            #####################################################\n",
    "            #### spp num. 0: genomic architecture parameters ####\n",
    "            #####################################################\n",
    "\n",
    "                'gen_arch': {\n",
    "                    #file defining custom genomic arch\n",
    "                    'gen_arch_file':            None,\n",
    "                    #num of loci\n",
    "                    'L':                        100,\n",
    "                    #starting allele frequency (None to draw freqs randomly)\n",
    "                    'start_p_fixed':            0.5,\n",
    "                    #whether to start neutral locus freqs at 0\n",
    "                    'start_neut_zero':          False,\n",
    "                    #genome-wide per-base neutral mut rate (0 to disable)\n",
    "                    'mu_neut':                  0,\n",
    "                    #genome-wide per-base deleterious mut rate (0 to disable)\n",
    "                    'mu_delet':                 0,\n",
    "                    #shape of distr of deleterious effect sizes\n",
    "                    'delet_alpha_distr_shape':  0.2,\n",
    "                    #scale of distr of deleterious effect sizes\n",
    "                    'delet_alpha_distr_scale':  0.2,\n",
    "                    #alpha of distr of recomb rates\n",
    "                    'r_distr_alpha':            None,\n",
    "                    #beta of distr of recomb rates\n",
    "                    'r_distr_beta':             None,\n",
    "                    #whether loci should be dominant (for allele '1')\n",
    "                    'dom':                      False,\n",
    "                    #whether to allow pleiotropy\n",
    "                    'pleiotropy':               False,\n",
    "                    #custom fn for drawing recomb rates\n",
    "                    'recomb_rate_custom_fn':    None,\n",
    "                    #number of recomb paths to hold in memory\n",
    "                    'n_recomb_paths_mem':       int(1e4),\n",
    "                    #total number of recomb paths to simulate\n",
    "                    'n_recomb_paths_tot':       int(1e5),\n",
    "                    #num of crossing-over events (i.e. recombs) to simulate\n",
    "                    'n_recomb_sims':            10_000,\n",
    "                    #whether to generate recombination paths at each timestep\n",
    "                    'allow_ad_hoc_recomb':      False,\n",
    "                    #whether to save mutation logs\n",
    "                    'mut_log':                  False,\n",
    "                    #whether to jitter recomb bps, to correctly track num_trees\n",
    "                    'jitter_breakpoints':       False,\n",
    "                    #whether to use tskit (to record full spatial pedigree)\n",
    "                    'use_tskit':                True,\n",
    "                    #time step interval for simplication of tskit tables\n",
    "                    'tskit_simp_interval':      100,\n",
    "\n",
    "                    }, # <END> 'gen_arch'\n",
    "\n",
    "\n",
    "                }, # <END> spp num. 0\n",
    "\n",
    "\n",
    "\n",
    "    #### NOTE: individual Species' sections can be copy-and-pasted (and\n",
    "    #### assigned distinct keys and names), to create additional Species.\n",
    "\n",
    "\n",
    "            }, # <END> 'species'\n",
    "\n",
    "        }, # <END> 'comm'\n",
    "\n",
    "\n",
    "###############################################################################\n",
    "\n",
    "###############\n",
    "#### MODEL ####\n",
    "###############\n",
    "    'model': {\n",
    "        #total Model runtime (in timesteps)\n",
    "        'T':            100,\n",
    "        #min burn-in runtime (in timesteps)\n",
    "        'burn_T':       30,\n",
    "        #seed number\n",
    "        'num':          None,\n",
    "\n",
    "\n",
    "        ###############################\n",
    "        #### iterations parameters ####\n",
    "        ###############################\n",
    "        'its': {\n",
    "            #num iterations\n",
    "            'n_its':            1,\n",
    "            #whether to randomize Landscape each iteration\n",
    "            'rand_landscape':   False,\n",
    "            #whether to randomize Community each iteration\n",
    "            'rand_comm':        False,\n",
    "            #whether to burn in each iteration\n",
    "            'repeat_burn':      False,\n",
    "            #whether to randomize GenomicArchitectures each iteration\n",
    "            'rand_genarch':     True, \n",
    "            }, # <END> 'iterations'\n",
    "\n",
    "\n",
    "\n",
    "        } # <END> 'model'\n",
    "\n",
    "    } # <END> params\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we'll do the same steps as before:\n",
    "- make the model\n",
    "- run it\n",
    "- plot the results\n",
    "\n",
    "To do that, go ahead and **run the next few cells**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = gnx.make_params_dict(params, 'heterogeneous_demo')\n",
    "mod = gnx.make_model(parameters=params, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mod.walk(T=10000, mode='burn')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mod.walk(T=10, mode='main')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mod.plot(lyr=0, spp=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- BEGIN QUESTION -->\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <b>Use this information to answer question 1 in the assignment.</b>\n",
    "    <br />    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-danger\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>PART 2: Environmental Change</h2></b> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we've seen a quick example of how Geonomics works, we'll make a more complicated model, then use the results to explore questions about gene flow, isolation, founder effects, and genetic divergence in species distributed across space.\n",
    "\n",
    "For example, **what would happen if a barrier were to arise on the landscape?** n the real world, this could happen in a number of ways.  For our mouse island, one example could be the construction of a highway.\n",
    "\n",
    "Let's simulate one to find out!\n",
    "\n",
    "This time, for the first 500 time steps, our species will evolve on a homogeneous landscape (just like our first model). We will make some plots to observe the model, we'll create another model with a homogeneous landscape layer, then run that model for 500 timesteps and observe how genetic relatedness changes over time.  *This will be equivalent to watching genetic evolution in the first model we made.* \n",
    "\n",
    "But then, a big central barrier will arise. We will see this barrier arise right after it arises, simulate another 500 time steps, then use the same plots to see how the barrier influenced genetic relatedness.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Concept check** *What do you expect to see after the appearance of the barrier?  For example, do you think individuals on opposite sides of the barrier will stay closely related after 500 generations?*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To start off, we need to create and save an array for our landscape layer. To do that, we'll use one of the helper functions that we loaded at the beginning. To do that, **run the next cell**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save a new file\n",
    "barr_lyr = add_vert_barrier(make_unif_array(40))\n",
    "plt.imshow(barr_lyr, cmap = 'RdBu_r')\n",
    "np.savetxt('barrier.txt', barr_lyr, fmt='%0.5f')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we'll create a new model, using extra parameters that set up a landscape change event that uses that file. \n",
    "\n",
    "Here are the key changes we've made in order to characterize that landscape change event (Reminder- you do not need to make any changes to the code, these definitions are to help you understand all the parameters of the model):\n",
    "\n",
    "- **dim** parameter: *We set the landscape dimensions to __(40, 40)__.*\n",
    "\n",
    "- **rast** parameter: *We set the landscape raster to a 40-cell by 40-cell uniform array of ones. We do this using a function we loaded at the beginning: type __`make_unif_array(40)`__.*\n",
    "\n",
    "- **change_rast** parameter: *We set the value of this parameter, in the landscape-change section, to __'./barrier.txt'__ (the barrier layer that we just created)*.\n",
    "\n",
    "- **start_t** parameter: *We set the landscape-change event to start at timestep __500__*.\n",
    "\n",
    "- **end_t** parameter: *We set the landscape-change event to end at timestep __600__*.\n",
    "\n",
    "- **n_steps** parameter: *To keep things simple, we set the number of stepwise-changes in the landscape-change event to just __2__*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# barrier_demo.py\n",
    "\n",
    "# This is a parameters file generated by Geonomics\n",
    "# (by the gnx.make_parameters_file() function).\n",
    "\n",
    "\n",
    "                   #   ::::::          :::    :: :::::::::::#\n",
    "             #::::::    ::::   :::      ::    :: :: ::::::::::: ::#\n",
    "          #:::::::::     ::            ::   :::::::::::::::::::::::::#\n",
    "        #::::::::::                      :::::::::: :::::: ::::::::  ::#\n",
    "      #  : ::::  ::                    ::::  : ::    :::::::: : ::  :    #\n",
    "     # GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #\n",
    "    # GG     EE    OO   OO  NNN  NN  OO   OO  MM   MM   II   CC     SS     #\n",
    "    # GG     EE   OO     OO NN N NN OO     OO MMM MMM   II   CC     SSSSSS #\n",
    "    # GG GGG EEEE OO     OO NN  NNN OO     OO MM M MM   II   CC         SS #\n",
    "    # GG   G EE    OO   OO  NN   NN  OO   OO  MM   MM   II   CC        SSS #\n",
    "     # GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #\n",
    "      #    : ::::::::               :::::::::: ::              ::  :   : #\n",
    "        #:    :::::                    :::::: :::             :::::::  #\n",
    "          #    :::                      :::::  ::              ::::: #\n",
    "             #  ::                      ::::                      #\n",
    "                   #                                        #\n",
    "                      #  :: ::    :::             #\n",
    "\n",
    "\n",
    "params = {\n",
    "###############################################################################\n",
    "\n",
    "###################\n",
    "#### LANDSCAPE ####\n",
    "###################\n",
    "    'landscape': {\n",
    "\n",
    "    ##############\n",
    "    #### main ####\n",
    "    ##############\n",
    "        'main': {\n",
    "            #x,y (a.k.a. j,i) dimensions of the Landscape\n",
    "            'dim':                      (40,40),\n",
    "            #x,y resolution of the Landscape\n",
    "            'res':                      (1,1),\n",
    "            #x,y coords of upper-left corner of the Landscape\n",
    "            'ulc':                      (0,0),\n",
    "            #projection of the Landscape\n",
    "            'prj':                      None,\n",
    "            }, # <END> 'main'\n",
    "\n",
    "    ################\n",
    "    #### layers ####\n",
    "    ################\n",
    "        'layers': {\n",
    "\n",
    "            #layer name (LAYER NAMES MUST BE UNIQUE!)\n",
    "            'lyr_0': {\n",
    "\n",
    "        #######################################\n",
    "        #### layer num. 0: init parameters ####\n",
    "        #######################################\n",
    "\n",
    "                #initiating parameters for this layer\n",
    "                'init': {\n",
    "\n",
    "                    #parameters for a 'defined'-type Layer\n",
    "                    'defined': {\n",
    "                        #raster to use for the Layer\n",
    "                        'rast':                   make_unif_array(40),\n",
    "                        #point coordinates\n",
    "                        'pts':                    None,\n",
    "                        #point values\n",
    "                        'vals':                   None,\n",
    "                        #interpolation method {None, 'linear', 'cubic',\n",
    "                        #'nearest'}\n",
    "                        'interp_method':          None,\n",
    "\n",
    "                        }, # <END> 'defined'\n",
    "\n",
    "                    }, # <END> 'init'\n",
    "\n",
    "            #########################################\n",
    "            #### layer num. 0: change parameters ####\n",
    "            #########################################\n",
    "\n",
    "                #landscape-change events for this Layer\n",
    "                'change': {\n",
    "\n",
    "                    0: {\n",
    "                        #array or file for final raster of event, or directory\n",
    "                        #of files for each stepwise change in event\n",
    "                        'change_rast':      './barrier.txt',\n",
    "                        #starting timestep of event\n",
    "                        'start_t':          500,\n",
    "                        #ending timestep of event\n",
    "                        'end_t':            600,\n",
    "                        #number of stepwise changes in event\n",
    "                        'n_steps':          1,\n",
    "                        }, # <END> event 0\n",
    "\n",
    "                    }, # <END> 'change'\n",
    "\n",
    "                }, # <END> layer num. 0\n",
    "\n",
    "\n",
    "\n",
    "    #### NOTE: Individual Layers' sections can be copy-and-pasted (and\n",
    "    #### assigned distinct keys and names), to create additional Layers.\n",
    "\n",
    "\n",
    "            } # <END> 'layers'\n",
    "\n",
    "        }, # <END> 'landscape'\n",
    "\n",
    "\n",
    "###############################################################################\n",
    "\n",
    "###################\n",
    "#### COMMUNITY ####\n",
    "###################\n",
    "    'comm': {\n",
    "\n",
    "        'species': {\n",
    "\n",
    "            #species name (SPECIES NAMES MUST BE UNIQUE!)\n",
    "            'spp_0': {\n",
    "\n",
    "            #####################################\n",
    "            #### spp num. 0: init parameters ####\n",
    "            #####################################\n",
    "\n",
    "                'init': {\n",
    "                    #starting number of individs\n",
    "                    'N':                250,\n",
    "                    #carrying-capacity Layer name\n",
    "                    'K_layer':          'lyr_0',\n",
    "                    #multiplicative factor for carrying-capacity layer\n",
    "                    'K_factor':         1,\n",
    "                    }, # <END> 'init'\n",
    "\n",
    "            #######################################\n",
    "            #### spp num. 0: mating parameters ####\n",
    "            #######################################\n",
    "\n",
    "                'mating'    : {\n",
    "                    #age(s) at sexual maturity (if tuple, female first)\n",
    "                    'repro_age':                0,\n",
    "                    #whether to assign sexes\n",
    "                    'sex':                      False,\n",
    "                    #ratio of males to females\n",
    "                    'sex_ratio':                1/1,\n",
    "                    #whether P(birth) should be weighted by parental dist\n",
    "                    'dist_weighted_birth':       False,\n",
    "                    #intrinsic growth rate\n",
    "                    'R':                        0.5,\n",
    "                    #intrinsic birth rate (MUST BE 0<=b<=1)\n",
    "                    'b':                        0.2,\n",
    "                    #expectation of distr of n offspring per mating pair\n",
    "                    'n_births_distr_lambda':    1,\n",
    "                    #whether n births should be fixed at n_births_dist_lambda\n",
    "                    'n_births_fixed':           True,\n",
    "                    #radius of mate-search area\n",
    "                    'mating_radius':            10,\n",
    "                    #whether individs should choose nearest neighs as mates\n",
    "                    'choose_nearest_mate':        False,\n",
    "                    #whether mate-choice should be inverse distance-weighted\n",
    "                    'inverse_dist_mating':      False,\n",
    "                    }, # <END> 'mating'\n",
    "\n",
    "            ##########################################\n",
    "            #### spp num. 0: mortality parameters ####\n",
    "            ##########################################\n",
    "\n",
    "                'mortality'     : {\n",
    "                    #maximum age\n",
    "                    'max_age':                      None,\n",
    "                    #min P(death) (MUST BE 0<=d_min<=1)\n",
    "                    'd_min':                        0,\n",
    "                    #max P(death) (MUST BE 0<=d_max<=1)\n",
    "                    'd_max':                        1,\n",
    "                    #width of window used to estimate local pop density\n",
    "                    'density_grid_window_width':    None,\n",
    "                    }, # <END> 'mortality'\n",
    "\n",
    "            #########################################\n",
    "            #### spp num. 0: movement parameters ####\n",
    "            #########################################\n",
    "\n",
    "                'movement': {\n",
    "                    #whether or not the species is mobile\n",
    "                    'move':                                 True,\n",
    "                    #mode of distr of movement direction\n",
    "                    'direction_distr_mu':                   0,\n",
    "                    #concentration of distr of movement direction\n",
    "                    'direction_distr_kappa':                0,\n",
    "                    #1st param of distr of movement distance\n",
    "                    'movement_distance_distr_param1':       0.5,\n",
    "                    #2nd param of distr of movement distance\n",
    "                    'movement_distance_distr_param2':       5e-8,\n",
    "                    #movement distance distr to use ('levy' or 'wald')\n",
    "                    'movement_distance_distr':              'levy',\n",
    "                    #1st param of distr of dispersal distance\n",
    "                    'dispersal_distance_distr_param1':      0,\n",
    "                    #2nd param of distr of dispersal distance\n",
    "                    'dispersal_distance_distr_param2':      5e-14,\n",
    "                    #dispersal distance distr to use ('levy' or 'wald')\n",
    "                    'dispersal_distance_distr':             'levy',\n",
    "                    'move_surf'     : {\n",
    "                        #move-surf Layer name\n",
    "                        'layer':                'lyr_0',\n",
    "                        #whether to use mixture distrs\n",
    "                        'mixture':              True,\n",
    "                        #concentration of distrs\n",
    "                        'vm_distr_kappa':       12,\n",
    "                        #length of approximation vectors for distrs\n",
    "                        'approx_len':           5000,\n",
    "                        }, # <END> 'move_surf'\n",
    "\n",
    "                    },    # <END> 'movement'\n",
    "\n",
    "\n",
    "            #####################################################\n",
    "            #### spp num. 0: genomic architecture parameters ####\n",
    "            #####################################################\n",
    "\n",
    "                'gen_arch': {\n",
    "                    #file defining custom genomic arch\n",
    "                    'gen_arch_file':            None,\n",
    "                    #num of loci\n",
    "                    'L':                        100,\n",
    "                    #starting allele frequency (None to draw freqs randomly)\n",
    "                    'start_p_fixed':            0.5,\n",
    "                    #whether to start neutral locus freqs at 0\n",
    "                    'start_neut_zero':          False,\n",
    "                    #genome-wide per-base neutral mut rate (0 to disable)\n",
    "                    'mu_neut':                  0,\n",
    "                    #genome-wide per-base deleterious mut rate (0 to disable)\n",
    "                    'mu_delet':                 0,\n",
    "                    #shape of distr of deleterious effect sizes\n",
    "                    'delet_alpha_distr_shape':  0.2,\n",
    "                    #scale of distr of deleterious effect sizes\n",
    "                    'delet_alpha_distr_scale':  0.2,\n",
    "                    #alpha of distr of recomb rates\n",
    "                    'r_distr_alpha':            None,\n",
    "                    #beta of distr of recomb rates\n",
    "                    'r_distr_beta':             None,\n",
    "                    #whether loci should be dominant (for allele '1')\n",
    "                    'dom':                      False,\n",
    "                    #whether to allow pleiotropy\n",
    "                    'pleiotropy':               False,\n",
    "                    #custom fn for drawing recomb rates\n",
    "                    'recomb_rate_custom_fn':    None,\n",
    "                    #number of recomb paths to hold in memory\n",
    "                    'n_recomb_paths_mem':       int(1e4),\n",
    "                    #total number of recomb paths to simulate\n",
    "                    'n_recomb_paths_tot':       int(1e5),\n",
    "                    #num of crossing-over events (i.e. recombs) to simulate\n",
    "                    'n_recomb_sims':            10_000,\n",
    "                    #whether to generate recombination paths at each timestep\n",
    "                    'allow_ad_hoc_recomb':      False,\n",
    "                    #whether to jitter recomb bps, to correctly track num_trees\n",
    "                    'jitter_breakpoints':       False,\n",
    "                    #whether to save mutation logs\n",
    "                    'mut_log':                  False,\n",
    "                    #whether to use tskit (to record full spatial pedigree)\n",
    "                    'use_tskit':                True,\n",
    "                    #time step interval for simplication of tskit tables\n",
    "                    'tskit_simp_interval':      100,\n",
    "\n",
    "\n",
    "                    }, # <END> 'gen_arch'\n",
    "\n",
    "\n",
    "                }, # <END> spp num. 0\n",
    "\n",
    "\n",
    "\n",
    "    #### NOTE: individual Species' sections can be copy-and-pasted (and\n",
    "    #### assigned distinct keys and names), to create additional Species.\n",
    "\n",
    "\n",
    "            }, # <END> 'species'\n",
    "\n",
    "        }, # <END> 'comm'\n",
    "\n",
    "\n",
    "###############################################################################\n",
    "\n",
    "###############\n",
    "#### MODEL ####\n",
    "###############\n",
    "    'model': {\n",
    "        #total Model runtime (in timesteps)\n",
    "        'T':            100,\n",
    "        #min burn-in runtime (in timesteps)\n",
    "        'burn_T':       30,\n",
    "        #seed number\n",
    "        'num':          None,\n",
    "\n",
    "        ###############################\n",
    "        #### iterations parameters ####\n",
    "        ###############################\n",
    "        'its': {\n",
    "            #num iterations\n",
    "            'n_its':            1,\n",
    "            #whether to randomize Landscape each iteration\n",
    "            'rand_landscape':   False,\n",
    "            #whether to randomize Community each iteration\n",
    "            'rand_comm':        False,\n",
    "            #whether to randomize GenomicArchitectures each iteration\n",
    "            'rand_genarch':     True,\n",
    "            #whether to burn in each iteration\n",
    "            'repeat_burn':      False,\n",
    "            }, # <END> 'iterations'\n",
    "\n",
    "\n",
    "\n",
    "        } # <END> 'model'\n",
    "\n",
    "    } # <END> params\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, just like before, we'll create the model, burn it in, then run it and observe.\n",
    "\n",
    "We'll start by just running the portion of the model **before** the landscape change event, to observe what happens during that time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make our params dict into a proper Geonomics ParamsDict object\n",
    "params = gnx.make_params_dict(params, 'barrier_demo')\n",
    "#then use it to make a model\n",
    "mod = gnx.make_model(parameters=params, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#First we have to run the model burn-in, as always\n",
    "\n",
    "#SCROLL DOWN TO OBSERVE BURN-IN AND SEE PLOT\n",
    "\n",
    "mod.walk(T=10000, mode='burn')\n",
    "#plot the resulting object\n",
    "mod.plot(spp=0, lyr=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Now we can run the main phase of the model for the first 500 timesteps.\n",
    "# NOTE: THIS WILL TAKE A FEW MINUTES TO RUN\n",
    "mod.walk(500, 'main')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then let's plot the individuals in \"genetic-relatedness space\",\n",
    "and also map them in geographic space but colored by their genetic relatedness.  Closely related individuals will be similarly colored."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#We'll use the genetic PCA functions we defined at the beginning.\n",
    "plot_PCA(mod)\n",
    "map_PCA(mod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- BEGIN QUESTION -->\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <b>Use this information to answer question 2 in the assignment.</b>\n",
    "    <br />   \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"font-size:120%\">\n",
    "    <b><h2>The effect of the barrier</h2></b> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will continue running this model. Over the next 50 timesteps, our landscape-change event will occur. The barrier will \"arise\" on the landscape, and will gradually reduce individuals' ability to move across it.\n",
    "\n",
    "We will plot our same genetic relatedness plots a few more times, so that we can see if the barrier has any effect on relatedness, and if so, how that effect unfolds over time.\n",
    "\n",
    "First, we'll run the model 10 more timesteps, the see the landscape right after the barrier has arisen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mod.walk(10, 'main')\n",
    "mod.plot(spp=0, lyr=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_PCA(mod)\n",
    "map_PCA(mod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we'll run the model for another 500 time steps, to give any effects ample time to play out. Then we'll plot genetic relatedness one last time."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then let's plot genetic relatedness again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# NOTE: THIS WILL TAKE A FEW MINUTES TO RUN\n",
    "mod.walk(500, 'main')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plot_PCA(mod)\n",
    "map_PCA(mod)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- BEGIN QUESTION -->\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "    <b> Use this information to answer questions 3-5 in the assignment.\".</b>\n",
    "    <br />\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook was developed by Drew Hart, Natalie Graham, Monica Wilkinson and Keeley Takimoto (2019) and contributed to by Anusha Bishop (2020)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
